Fix issue with vertical lines not visible in the elevation profile tool

ptitjano opened the pull request:
## Description

This PR reverts commit ce96dee6b517bc19a84c9394e48d6cb30e80c1f7.

This commit was introduced to gain some performance when the
`tolerance` is set by using `request.setDistanceWithin` instead of
`request.setFilterRect`. However, in that case, the providers either
use `GEOS::distance` or `GEOS::distanceWithin` which is not able to
handle the Z component of a Geometry.
For example, a purely vertical line, i.e.: `LineString(X Y Z1, X Y
Z2)` will always be to an infinite distance of any geometry according
to GEOS.

In practice, this means that a purely vertical line will never be
visible in the elevation profile when the tolerance is set. Indeed,
the provider will always discard such geometry because of the usage
of `setDistanceWithin`. This issue is fixed by always using
`setFilterRect`. This might introduce a performance penalty but this
ensures that the result is always correct.

## Update October 10

```
`GEOSPreparedDistance_r` is not able to properly handle a purely
vertical line (LineString Z((X Y Z1, X Y Z2, ..., X Y Zn))). In that
case, it always `inf`. However, `GEOSDistance_r` works.

This issue is fixed by checking if one of the geometry is vertical 3d
line by introducing `isZVerticalLine`. This function checks if a
geometry is a 3d line. If that's the case, it loops through the points
of the line. If two points do not have the same X Y corrdinates, it
means that the line is not purely vertical.

If `geom` is a vertical line, it is replaced by its first point.
It the prepared geometry is a vertical line, fallback to the non
prepared case.
```

ptitjano committed:
test_qgsvectorlayerprofilegenerator: Add a test for purely vertical lines

See the two previous commits

nyalldawson commented:
> However, in that case, the providers either
use GEOS::distance or GEOS::distanceWithin which is not able to
handle the Z component of a Geometry.
For example, a purely vertical line, i.e.: LineString(X Y Z1, X Y Z2) will always be to an infinite distance of any geometry according
to GEOS.

I think this special case should be handled in the iterator (or geometry) classes then, not specifically the elevation profile code

ptitjano commented:
> > However, in that case, the providers either
> > use GEOS::distance or GEOS::distanceWithin which is not able to
> > handle the Z component of a Geometry.
> > For example, a purely vertical line, i.e.: LineString(X Y Z1, X Y Z2) will always be to an infinite distance of any geometry according
> > to GEOS.
> 
> I think this special case should be handled in the iterator (or geometry) classes then, not specifically the elevation profile code

Thanks for you message. However, I don't understand it. Do you mean that the iterator code of the providers should be changed? Or something else?

nyalldawson commented:
. Do you mean that the iterator code of the providers should be changed?

I mean if the setDistanceWithin method for QgsFeatureRequest can't handle this situation, then it needs to be fixed in the iterator code. Otherwise other users of setDistanceWithin will remain broken. 

So maybe the fix needs to apply in QgsGeos::distance / QgsGeos::distanceWithin instead, eg by detecting a vertical line and using a GEOS point geometry there instead...

ptitjano committed:
qgsgeos: Properly handle Z vertical lines in distance

`GEOSPreparedDistance_r` is not able to properly handle a purely
vertical line (LineString Z((X Y Z1, X Y Z2, ..., X Y Zn))). In that
case, it always `inf`. However, `GEOSDistance_r` works.

This issue is fixed by checking if one of the geometry is vertical 3d
line by introducing `isZVerticalLine`. This function checks if a
geometry is a 3d line. If that's the case, it loops through the points
of the line. If two points do not have the same X Y corrdinates, it
means that the line is not purely vertical.

If `geom` is a vertical line, it is replaced by its first point.
It the prepared geometry is a vertical line, fallback to the non
prepared case.

ptitjano committed:
Properly handle Z vertical lines in distanceWithin

This is similar to the change done on `QgsGeos::distance` in the
previous commit.

ptitjano commented:
> . Do you mean that the iterator code of the providers should be changed?
> 
> I mean if the setDistanceWithin method for QgsFeatureRequest can't handle this situation, then it needs to be fixed in the iterator code. Otherwise other users of setDistanceWithin will remain broken.
> 
> So maybe the fix needs to apply in QgsGeos::distance / QgsGeos::distanceWithin instead, eg by detecting a vertical line and using a GEOS point geometry there instead...

Thanks for the suggestion. I have completely rewritten the PR to handle this is in `QgsGeos`.

nyalldawson commented on src/core/geometry/qgsgeos.cpp:
```diff
@@ -540,6 +541,44 @@ QgsAbstractGeometry *QgsGeos::symDifference( const QgsAbstractGeometry *geom, QS
+    return false;
+  }
+
+  std::unique_ptr<QgsLineString> line( qgsgeometry_cast<QgsLineString *>( geom->clone() ) );
```
Why a clone vs just casting in place?

ptitjano replied:
This was la left-over from a previous iteration of the change. Fixed.

nyalldawson commented on src/core/geometry/qgsgeos.cpp:
```diff
@@ -540,6 +541,44 @@ QgsAbstractGeometry *QgsGeos::symDifference( const QgsAbstractGeometry *geom, QS
+
+  std::unique_ptr<QgsLineString> line( qgsgeometry_cast<QgsLineString *>( geom->clone() ) );
+  bool isVertical = true;
+  if ( line )
```
```suggestion
  if ( const QgsLineString* line = qgsgeometry_cast<const QgsLineString *>( geom ) )
```

ptitjano replied:
Changed

nyalldawson commented on src/core/geometry/qgsgeos.cpp:
```diff
@@ -540,6 +541,44 @@ QgsAbstractGeometry *QgsGeos::symDifference( const QgsAbstractGeometry *geom, QS
+    const double sqrTolerance = tolerance * tolerance;
+    for ( int iVert = nrPoints - 1, jVert = 0; jVert < nrPoints; iVert = jVert++ )
+    {
+      const QgsPoint pi = geom->vertexAt( QgsVertexId( 0, 0, iVert ) );
```
vertexAt is a slow approach to iterate through vertices -- better is something like https://github.com/qgis/QGIS/blob/master/src/core/symbology/qgssymbollayerutils.cpp#L4452 .

ptitjano replied:
This should be good now.

nyalldawson commented on src/core/geometry/qgsgeos.cpp:
```diff
@@ -548,7 +587,21 @@ double QgsGeos::distance( const QgsAbstractGeometry *geom, QString *errorMsg ) c
+  // In that case, replace `geom` by a single point.
+  // However, GEOSDistance_r works.
+  std::unique_ptr< QgsAbstractGeometry > preparedGeom;
+  if ( mGeosPrepared && isZVerticalLine( geom ) )
```
There's an issue here if geom is a multilinestring containing a single line. Best way to handle this is:

```suggestion
  if ( mGeosPrepared && isZVerticalLine( geom->simplifiedTypeRef() ) )
```

ptitjano replied:
Done.

nyalldawson commented on src/core/geometry/qgsgeos.cpp:
```diff
@@ -548,7 +587,21 @@ double QgsGeos::distance( const QgsAbstractGeometry *geom, QString *errorMsg ) c
+  }
+  else
+  {
+    preparedGeom.reset( geom->clone() );
```
This is inefficient -- it's forcing an unnecessary clone in the standard (non vertical line) case. 

Move `geos::unique_ptr otherGeosGeom'` to before this check, and avoid setting preparedGeom when we don't need it.

ptitjano replied:
Done.

nyalldawson commented on src/core/geometry/qgsgeos.cpp:
```diff
@@ -607,7 +660,17 @@ bool QgsGeos::distanceWithin( const QgsAbstractGeometry *geom, double maxdist, Q
 
-  geos::unique_ptr otherGeosGeom( asGeos( geom, mPrecision ) );
+  std::unique_ptr< QgsAbstractGeometry > preparedGeom;
+  if ( mGeosPrepared && isZVerticalLine( geom ) )
```
Same notes as above -- handle multilinestrings, avoid unwanted clone

nyalldawson commented:
Thanks @ptitjano , this approach looks much better to me! Just be aware of introducing any inefficiencies in these classes (like clones which aren't 100% required), as this is all performance critical code which needs to be as optimised as possible

ptitjano commented:
> Thanks @ptitjano , this approach looks much better to me! Just be aware of introducing any inefficiencies in these classes (like clones which aren't 100% required), as this is all performance critical code which needs to be as optimised as possible

Thanks. This should be good now.

nyalldawson commented on src/core/geometry/qgsgeos.cpp:
```diff
@@ -548,7 +586,21 @@ double QgsGeos::distance( const QgsAbstractGeometry *geom, QString *errorMsg ) c
+  // However, GEOSDistance_r works.
+  if ( mGeosPrepared && isZVerticalLine( geom->simplifiedTypeRef() ) )
+  {
+    otherGeosGeom = asGeos( geom->vertexAt( QgsVertexId( 0, 0, 0 ) ).clone(), mPrecision );
```
asGeos doesn't take ownership, so this leaks the cloned point

ptitjano replied:
Fixed.
